Feat: Initial Implementation of Grayskull Secret Management Service

imanadi opened the pull request:
### Overview
This PR establishes the foundational MVP implementation for Grayskull, a secret management service. It introduces a modular, extensible architecture with Spring Boot-based REST APIs, persistence-agnostic SPI interfaces, MongoDB-backed storage, fine-grained authorization, comprehensive auditing, and containerized local development setup.

---

### REST API Layer (`SecretController`)

**Endpoints:**
```
GET    /v1/projects/{projectId}/secrets
       - List secrets with pagination (offset/limit)
       - Secured: secrets.list permission
       - Returns latest version of each secret

POST   /v1/projects/{projectId}/secrets
       - Create new secret with initial data (version 1)
       - Secured: secrets.create permission
       - Supports SELF provider with public/private parts

GET    /v1/projects/{projectId}/secrets/{secretName}
       - Read secret metadata (name, version, timestamps, state)
       - Secured: secrets.read.metadata permission
       - Excludes sensitive data values

GET    /v1/projects/{projectId}/secrets/{secretName}/data
       - Read decrypted secret value
       - Secured: secrets.read.value permission
       - Returns both public and private parts

POST   /v1/projects/{projectId}/secrets/{secretName}/data
       - Upgrade secret data (creates new version)
       - Secured: secrets.update permission
       - Immutable versioning: previous versions preserved

DELETE /v1/projects/{projectId}/secrets/{secretName}
       - Soft-delete (marks as DISABLED)
       - Secured: secrets.delete permission
       - Audit trail maintained

GET    /v1/projects/{projectId}/secrets/{secretName}/versions/{version}
       - Admin API: Retrieve specific version
       - Secured: secrets.read.version permission
       - Supports state filtering (ACTIVE/DISABLED/DELETED)
```

**Request/Response Handling:**
- Jakarta Bean Validation for input sanitization (`@Valid`, `@NotBlank`, `@Size`, constraint ranges)
- `ResponseTemplate<T>` wrapper with success/error indicators and messages

---

### Architecture & Module Structure

#### 1. SPI Module (Service Provider Interface)
**Purpose:** Defines persistence-agnostic contracts and domain models for extensibility.

**Key Components:**
- **Interfaces:**
  - `GrayskullAuthorizationProvider`: Pluggable authorization provider SPI
  - `SecretRepository`, `SecretDataRepository`, `ProjectRepository`, `AuditEntryRepository`: Persistence-agnostic repository contracts
  
- **Domain Models:** `Secret`, `SecretData`, `Project`, `AuditEntry` with lifecycle states (`ACTIVE`, `DISABLED`, `DELETED`)

- **Authorization Context:** `AuthorizationContext` for project and secret-level permission checks

**Design Philosophy:** Zero external dependencies (only provided-scope framework deps), enabling future implementations (e.g., different databases, cloud provider integrations)

#### 2. Server Module
**Core Spring Boot application providing business logic and persistence.**

##### Authorization Framework
**Implementation:**
- **`GrayskullSecurity`**: Security facade bean for `@PreAuthorize` expressions
  - `hasPermission(projectId, action)` - Project-level checks (e.g., create, list)
  - `hasPermission(projectId, secretName, action)` - Secret-level checks (e.g., read, update)
  
- **`SimpleAuthorizationProvider`**: Configuration-driven rule engine with wildcard support
  - Rules defined in `application.yml` under `grayskull.authorization.rules`
  - Supports user, project, and secret-level matching with `*` wildcards
  
- **`AuthorizationProperties`**: Type-safe configuration binding for authorization rules

**Action Definitions (`GrayskullActions`):** 
- `secrets.create`, `secrets.list`, `secrets.read.metadata`, `secrets.read.value`, `secrets.read.version`, `secrets.update`, `secrets.delete`

**Authorization Logic:**
- Non-existent projects → Transient project with default KMS key for create operations
- Non-existent secrets → Fallback to project-level rules (allows permission checks before resource exists)
- Existing resources → Full secret-level permission evaluation

##### Audit Framework (AOP-based)
**Implementation:**
- **`@Audit`** annotation with `AuditAction` enum for marking auditable operations
- **`AuditAspect`**: Spring AOP aspect with `@AfterReturning` advice
  
**Behavior:**
- Audits only successful CUD (Create, Update, Delete) operations
- Captures: `userId`, `projectId`, `resourceType`, `resourceName`, `resourceVersion`, `action`, `timestamp`
- Metadata includes sanitized request/response payloads
- **Masking:** `@AuditMask` annotation + custom Jackson serializer for PII/sensitive data
- Transaction-aligned: Failures in audit serialization rollback the entire operation

##### Service Layer
**`SecretService` / `SecretServiceImpl`:**
- Secret lifecycle management with versioning (immutable secret data)
- Soft-delete implementation (marks secrets as `DISABLED`)
- KMS key inheritance from project configuration
- Concurrency: Optimistic locking via MongoDB version field

##### Exception Handling
**`GlobalExceptionHandler`** (`@RestControllerAdvice`):
- Centralized exception handling with consistent error responses
- Validation errors return detailed `Violation` objects
- Custom exceptions mapped to appropriate HTTP status codes

##### Persistence Layer
**MongoDB Implementation:**
- Entity classes: `SecretEntity`, `SecretDataEntity`, `ProjectEntity`, `AuditEntryEntity`
- Repository implementations: Delegates to `MongoRepository` interfaces
- Timestamp generation delegated to DB (`@CreatedDate`, `@LastModifiedDate`)
- Separate collections for metadata and data (optimized reads)

##### Configuration
- **`CryptoConfig`**: Manages encryption keys (bound to `grayskull.crypto.keys.*`)
- **`KmsConfig`**: Default KMS key configuration
- **`SecurityConfiguration`**: Spring Security with Basic Auth
- **`MongoBeanConfiguration`**: Auditing enablement, custom converters

#### 3. Setup Module
**Local Development Environment:**
- Docker Compose setup for 3-node MongoDB replica set (ports 37017-37019)
- Authentication with keyfile-based security
- Scripts: `setup.sh` (automated initialization), `teardown.sh` (cleanup)
- Environment configuration via `.env` file

#### 4. Simple-App Module
**Purpose:** Example application demonstrating Grayskull integration as a library.

---

### Testing Strategy

**Framework:** JUnit 5 + Spring Boot Test + Testcontainers

**`BaseIntegrationTest`:**
- Spins up ephemeral MongoDB container per test run
- `@DynamicPropertySource` for runtime configuration
- Provides helper methods for API operations with user context
- 12 test classes covering controllers, mappers, models, and authorization

**Test Coverage:**
- **API Tests:** `SecretControllerIT` - CRUD operations, pagination, versioning, authorization scenarios
- **Unit Tests:** `SecretMapperTest`, `SecretTest`, `SecretDataTest`, `AuditMaskingSerializerTest`
- **Authorization Tests:** Multi-role scenarios (admin, editor, viewer) with `@PreAuthorize` validation

**Test Configuration (`test/resources/application.yml`):**
- Predefined authorization rules for consistent test behavior
- User roles: `admin` (full access), `editor` (create/read/update), `viewer` (read-only)

---

### Build & Tooling

**Maven Multi-Module:**
- Parent POM with centralized dependency management
- Spring Boot 3.5.0, Java 21
- Dependencies: Spring Data MongoDB, Spring Security, MapStruct, Lombok, Jackson, Testcontainers

alekhrj commented on models/src/main/java/com/flipkart/grayskull/models/db/Project.java:
```diff
@@ -0,0 +1,37 @@
+     * The organization or tenant this project belongs to (e.g., "FLIPKART", "FLIPKART/CT").
+     * This is crucial for multi-tenancy, ensuring data isolation.
+     */
+    private String org;
```
Please check if this makes sense to have Org as a separate entity itself. There will be policies associated with the org.

imanadi replied:
Resolving. The org construct was removed.

alekhrj commented on models/src/main/java/com/flipkart/grayskull/models/dto/request/CreateSecretRequest.java:
```diff
@@ -0,0 +1,58 @@
+    @Data
+    @NoArgsConstructor
+    @AllArgsConstructor
+    public static class SecretDataPayload {
```
Do we need a static class here? UpdateSecretDataRequest class has the same variable.

imanadi replied:
Acknowledged, I will move it to a separate class.

alekhrj commented on server/src/main/java/com/flipkart/grayskull/controllers/AdminController.java:
```diff
@@ -0,0 +1,35 @@
+@RequestMapping("/v1/admin")
+@AllArgsConstructor
+@Validated
+public class AdminController {
```
lets double check if we need separate endpoints for this API. can be part of SecretContoller itself. we can figure out a way to do admin authorisation on those APIs.

imanadi replied:
Have consolidated the Admin and Secret controller into one based on discussion.

alekhrj commented on server/src/main/java/com/flipkart/grayskull/controllers/SecretController.java:
```diff
@@ -0,0 +1,76 @@
+    @Operation(summary = "Lists secrets for a given project with pagination. Always returns the latest version of the secret.")
+    @GetMapping
+    @PreAuthorize("hasPermission(#projectId, T(com.flipkart.grayskull.models.authz.GrayskullActions).LIST_SECRETS)")
+    public ListSecretsResponse listSecrets(@PathVariable("projectId") @NotBlank @Size(max = 255) String projectId,
```
`
class Response<T> {
    T data;
     String message;
    String errorCode;
}
`
we can go with standard template for all type of responses, including error responses. 

easier on client integration, especially UI.

alekhrj commented on server/src/main/java/com/flipkart/grayskull/authz/GrayskullPermissionEvaluator.java:
```diff
@@ -0,0 +1,36 @@
+ */
+@Component
+@RequiredArgsConstructor
+public class GrayskullPermissionEvaluator implements PermissionEvaluator {
```
we don't need to mention (or prefix) Grayskull everywhere. com.flipkart.graykull as group id should be enough I think.

alekhrj commented on server/src/main/java/com/flipkart/grayskull/service/utils/SecretEncryptionUtil.java:
```diff
@@ -0,0 +1,30 @@
+
+@Component
+@RequiredArgsConstructor
+public class SecretEncryptionUtil {
```
looks similar to SecretEncryptionHelper.java.

imanadi replied:
Removed the duplicate.

alekhrj commented on server/src/main/java/com/flipkart/grayskull/repositories/AuditEntryRepository.java:
```diff
@@ -0,0 +1,13 @@
+ * Repository interface for {@link AuditEntry} entities.
+ */
+@Repository
+public interface AuditEntryRepository extends MongoRepository<AuditEntry, String> {
```
we can define the interface on DBAdapter. maybe split it across multiple interfaces, one for each entity.

imanadi replied:
We have moved the DB repositories interface to spi.

ssrangisetti commented on models/pom.xml:
```diff
@@ -0,0 +1,61 @@
+        <dependency>
+            <groupId>org.projectlombok</groupId>
+            <artifactId>lombok</artifactId>
+            <version>1.18.38</version>
```
scope for lombok should be provided

ssrangisetti commented on models/src/main/java/com/flipkart/grayskull/models/authz/GrayskullActions.java:
```diff
@@ -0,0 +1,52 @@
+ * These actions are used to enforce permissions for various operations.
+ */
+@UtilityClass
+public class GrayskullActions {
```
check if an enum makes sense here

imanadi replied:
Resolving the comment, Implemented changes in commit: https://github.com/flipkart-incubator/grayskull/pull/3/commits/d9cf1a9062b842ffaa8e3d8393eff2e92ecac689

TanyaSravan commented on models/src/main/java/com/flipkart/grayskull/models/db/SecretData.java:
```diff
@@ -0,0 +1,70 @@
+     * The identifier of the Key Management Service (KMS) key used for encrypting and decrypting the {@code privatePart}.
+     */
+    private String kmsKeyId;
+
```
Should we allow keys on a secret version level? Wont it unnecessarily bloat the number of keys a kms would need to manage

imanadi replied:
It can increase the number of keys KMS has to manage, but as of now the implementation inherits a global key and prevents this. 
Additionally the support to change key would be added post MVP.

imanadi commented:
## High-Level Changes from Recent PR Review

Hello everyone,

Based on the recent PR review, I've implemented several significant changes to the project. Here are the high-level updates:

* **Repository Decoupling:** I've separated repository concerns by creating **repository interfaces** in the `spi` module and placing **DB-specific implementations** (for MongoDB) in the `server` module. This significantly decouples our data access layer.
* **Audit Masking:** Sensitive information in **audit logs is now masked** to enhance security and compliance.
* **Database Timestamp Delegation:** Timestamp generation (for creation, updates, and audit timestamps) has been **delegated to MongoDB's primary instance** to ensure consistency across records.
* **Streamlined KMS Key ID Inheritance:** A **globally configurable project** with a global KMS Key ID can now be set in the configuration. The project structure has been updated to facilitate easier internal code creation around this.
* **Soft Delete for Secrets:** Secrets and their versions are no longer permanently deleted. Instead, they are set to a **disabled state**, making them unavailable to regular users but still accessible to administrators.
* **Consistent Response Template:** A **standardized response template** has been introduced to ensure a consistent output format, aligning with other internal repositories.
* **Admin API Migration:** The **Admin API has been migrated to the Secret controller** for better organization and management.
* **Improved Authorization Context:** I've introduced a **parameter for `Authorization Context` in the authorization SPI** and enhanced the **SPEL expressions for the controller**, significantly improving readability.

imanadi commented on models/src/main/java/com/flipkart/grayskull/models/db/AuditEntry.java:
```diff
@@ -0,0 +1,70 @@
+    /**
+     * The outcome status of the action (e.g., "SUCCESS", "FAILURE", "PENDING").
+     */
+    private String status;
```
This is done. Marking as resolved.

imanadi commented on setup/README.md:
```diff
@@ -0,0 +1,55 @@
+# Grayskull Local Development Setup
```
Sure, I’ll move this to the global README once we define one.
As of now, we haven’t started writing the global README, but I think we should have one.
I’ll need to think about what all can be included in its scope.

Keeping this as is until finalization.

imanadi commented on models/src/main/java/com/flipkart/grayskull/models/dto/response/CreateSecretResponse.java:
```diff
@@ -0,0 +1,22 @@
+@JsonInclude(JsonInclude.Include.NON_NULL)
+public class CreateSecretResponse {
+    String projectId;
+    String name;
```
Enhanced models documentation with examples and consistency in the latest commit: https://github.com/flipkart-incubator/grayskull/pull/3/commits/96775b8f312e34806302299d49979c0cf42d423a

imanadi commented on models/src/main/java/com/flipkart/grayskull/models/db/Secret.java:
```diff
@@ -0,0 +1,120 @@
+     * The latest (current) {@link SecretData} associated with this secret.
+     * This field might not always be populated depending on the query to avoid fetching sensitive data unnecessarily.
+     */
+    @Transient
```
Thanks for the review! "@Transient" is actually correct here. Let me clarify:

**Intent:** This field temporarily combines Secret metadata with SecretData for response DTOs. We store them separately for security/performance but sometimes need to combine them.

**Why @Transient works:** This is Spring Data's @Transient (not Java's transient keyword). It tells MongoDB "don't persist this field" - which is exactly what we want since SecretData is stored in a separate collection.

**Addressing your concerns:**
- "Not persisted": Correct and intentional - avoids data duplication
- "Not inspected by mapping framework": Actually MapStruct can access @Transient fields fine (see SecretMapper.secretToCreateSecretResponse())

**Usage:**
```java
// In createSecret():
savedSecret.setData(secretData);  // Temporary assignment
return mapper.secretToCreateSecretResponse(savedSecret);  // Mapper uses it
```

The annotation prevents persistence while allowing temporary data combination for clean response mapping.

imanadi commented on models/src/main/java/com/flipkart/grayskull/models/db/SecretData.java:
```diff
@@ -0,0 +1,70 @@
+     * The public part of the secret, which can be exposed without compromising security.
+     * Optional, and its usage depends on the secret type.
+     */
+    private String publicPart;
```
The terminology was reviewed and approved during the HLD and LLD phases, as well as the API contract review.
The service has been implemented using this approved naming convention—for example, publicPart.

imanadi commented on models/src/main/java/com/flipkart/grayskull/models/authz/GrayskullActions.java:
```diff
@@ -0,0 +1,52 @@
+    /**
+     * Permission to list the secrets within a project.
+     */
+    public static final String LIST_SECRETS = "LIST_SECRETS";
```
Could you please elaborate on why that would be necessary and what additional benefits it would provide?
The current construct appears simpler and functions as expected.

grssam replied:
It's basically unstructured as of now. not even an enum. enforcing any structure inside string values would be difficult to maintain.. why exactly LIST_SECRETS? why not LIST-SECRETS or listSecrets? why the word "list"? will LIST_SECRETS be diffrent from GET_SECRET (say, for single secret) - from authz perspective? the core authz actions should be based on CRUD actions .. list isn't one of them.

imanadi replied:
Acknowledged, I will have a look, redesign and propose.

ssrangisetti commented on server/src/main/java/com/flipkart/grayskull/controllers/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,136 @@
+     */
+    @ExceptionHandler(Exception.class)
+    public ResponseEntity<ResponseTemplate<Void>> handleAllExceptions(Exception ex, WebRequest request) {
+        log.error("An unexpected error occurred", ex);
```
reduce the number of log lines for unhandled exception. otherwise someone can bombard this which increase our logs

grssam commented on server/src/main/java/com/flipkart/grayskull/mappers/SecretMapper.java:
```diff
@@ -0,0 +1,102 @@
+ */
+@Mapper(componentModel = "spring", unmappedTargetPolicy = ReportingPolicy.IGNORE,
+        imports = {UUID.class, Instant.class, SecretState.class})
+public interface SecretMapper {
```
is this class tested? please add tests for each type of mapping with various inputs .. null, malformed, partial null..

grssam commented on server/src/main/java/com/flipkart/grayskull/service/implementations/SecretServiceImpl.java:
```diff
@@ -0,0 +1,245 @@
+        Secret secret = findActiveSecretOrThrow(projectId, secretName);
+
+        String keyId = resolveKmsKeyId(projectId);
+        int newVersion = secret.getCurrentDataVersion() + 1;
```
this is bound to give race conditions. multiple upgrade calls hitting same pod/different pod.. will lead to multiple new secret data versions with same version id as they reference the version in db and add 1 to it.
that's why my other comment on using mongo based auto increment

imanadi replied:
You can check the file: models-db/src/main/java/com/flipkart/grayskull/models/db/Secret.java In the commit: https://github.com/flipkart-incubator/grayskull/commit/b2344162972304c3d97a0e73e48ac73548cd8709

Explaination:

The race condition is prevented by the `@Version` annotation I added in commit b234416. 

**How Optimistic Locking Prevents the Race Condition:**

When two concurrent requests try to upgrade the same secret:

1. Thread A reads Secret: `currentDataVersion=1`, `version=10`
2. Thread B reads Secret: `currentDataVersion=1`, `version=10` (same state)
3. Both calculate: `newVersion = 1 + 1 = 2`
4. Thread A calls `secretRepository.save(secret)`:
   - Spring Data checks: entity `version=10` matches DB `version=10` ✓
   - Updates DB: `currentDataVersion=2`, `version=11`
   - Creates SecretData with version 2
   - Thread A succeeds ✓
5. Thread B calls `secretRepository.save(secret)`:
   - Spring Data checks: entity `version=10` vs DB `version=11` ✗
   - Throws `OptimisticLockingFailureException`
   - No SecretData created
   - Thread B fails ✓

**Result:** Only one thread succeeds, no duplicate SecretData is created.

The `+1` operation is safe because the `@Version` field ensures only one thread can successfully commit each increment. The Secret entity is updated BEFORE creating SecretData to fail fast and prevent orphaned data.

**Note:** This is standard optimistic locking behavior. Clients should implement retry logic for `OptimisticLockingFailureException`, which is a common pattern for handling concurrent modifications.

imanadi replied:
Yes, optimistic locking works correctly across different pods/instances. **To directly answer your three questions:**

1. ✅ **Does it work across different pods?** Yes
2. ✅ **Will the update DB step still fail?** Yes, the second pod's update will fail
3. ✅ **Is there a unique constraint on the version field?** No unique constraint needed

**How It Works Across Pods:**

The `@Version` annotation creates **database-level concurrency control**, not application-level. When `secretRepository.save(secret)` is called:

1. **Pod A reads Secret:** `currentDataVersion=1`, `version=10`
2. **Pod B reads Secret:** `currentDataVersion=1`, `version=10` (same state from DB)
3. Both calculate: `newVersion = 1 + 1 = 2`
4. **Pod A saves to MongoDB:**
   - Spring Data generates update with query filter: `{ _id: "...", version: 10 }`
   - And update operation: `{ $set: { currentDataVersion: 2, ... }, $inc: { version: 1 } }`
   - MongoDB atomically checks version=10 matches and updates to version=11 ✓
   - **Result:** 1 document updated
   - Pod A succeeds ✓
5. **Pod B tries to save to MongoDB:**
   - Spring Data generates update with query filter: `{ _id: "...", version: 10 }`
   - MongoDB searches for document with _id AND version=10
   - **Result:** 0 documents match (version is now 11)
   - Spring Data detects 0 documents updated
   - Spring Data throws `OptimisticLockingFailureException`
   - Pod B fails ✓

**Why No Unique Constraint Is Needed:**

The version field doesn't need a unique constraint because:
- Spring Data includes the **current version in the query filter** (WHERE clause equivalent)
- MongoDB's update operation is **atomic at the document level**
- If the version doesn't match, MongoDB simply updates 0 documents (no error at DB level)
- Spring Data detects the 0-document update and throws the exception
- This is how optimistic locking works: **compare-and-swap at the database level**

A unique constraint wouldn't help here because:
- Multiple documents can have the same version number (different secrets)
- Each document's version only needs to increment correctly for that specific document
- The atomicity comes from MongoDB's single-document ACID guarantees, not from constraints

**Key Points:**
- ✅ The version check happens **in MongoDB**, not in application memory
- ✅ Works identically whether requests hit the same pod or different pods
- ✅ The database is the single source of truth for version conflicts
- ✅ This is standard optimistic locking (compare-and-swap) behavior across distributed systems
- ⚠️ Clients should implement retry logic for concurrent modifications

**References:**
- `SecretEntity.java` line 40-44: `@Version` annotation on `getVersion()`
- `Secret.java` line 91-95: Version field documentation

grssam commented on models-api/src/main/java/com/flipkart/grayskull/models/dto/response/ResponseTemplate.java:
```diff
@@ -0,0 +1,66 @@
+     * @return A ResponseTemplate instance for a successful operation with data.
+     */
+    public static <T> ResponseTemplate<T> success(T data, String message) {
+        return new ResponseTemplate<>(data, message, null);
```
A success response shouldn't have a message. just the actual response and any meta.

imanadi replied:
Improved the Response template in the latest commit: https://github.com/flipkart-incubator/grayskull/pull/3/commits/7fc33ee53b867c82c04ef7f1015ab97db6035303

grssam reviewed (changes requested):
Few high level things:

- Please work on adding code coverage check
- you need to add many more tests for covering other kind of classes. you will get an idea once you add coverage report
- Is there metrics integration done already? which all components in the code are inherently generating metrics?

**simple-app and spi are pending**

grssam commented on spi/pom.xml:
```diff
@@ -18,6 +18,21 @@
+            <version>${project.version}</version>
+        </dependency>
+        <dependency>
+            <groupId>org.springframework.data</groupId>
+            <artifactId>spring-data-commons</artifactId>
+            <scope>provided</scope>
+        </dependency>
```
@imanadi - are you checking the AI reviews at all? we should disable these reviews if we are not finding them useful or not taking the input.

grssam commented on spi/pom.xml:
```diff
@@ -28,6 +42,29 @@
+                        <path>
+                            <groupId>org.projectlombok</groupId>
+                            <artifactId>lombok</artifactId>
+                            <version>${lombok.version}</version>
```
should not have any versions in child poms unless an override. Please remove. Please use parent-child relation and dependency management techniques

imanadi replied:
This suggestion is implemented in the current proposed code. Resolving the comment.

grssam commented on spi/src/main/java/com/flipkart/grayskull/spi/authz/AuthorizationContext.java:
```diff
@@ -0,0 +1,75 @@
+    private final Authentication authentication;
+
+    /** The project scope for the authorization check. */
+    private final Project project;
```
philosophical question - why should grayskull.spi.authz package depend on grayskul.models packages? Should we remove the direct reference of these objects and make them primitive data types? you only really need the projetId string and secret name string here

grssam commented on spi/src/main/java/com/flipkart/grayskull/spi/repositories/SecretDataRepository.java:
```diff
@@ -0,0 +1,23 @@
+     * @param dataVersion The version of the data to retrieve.
+     * @return An {@link Optional} containing the secret data if found.
+     */
+    Optional<SecretData> findBySecretIdAndDataVersion(String secretId, long dataVersion);
```
why `find` and not `get` ?

imanadi replied:
Resolving the comment, Implemented changes in commit: https://github.com/flipkart-incubator/grayskull/pull/3/commits/d9cf1a9062b842ffaa8e3d8393eff2e92ecac689

grssam commented on spi/src/main/java/com/flipkart/grayskull/spi/repositories/SecretDataRepository.java:
where is the default implementation for this class and other spi interfaces you are adding in this PR?

grssam reviewed (changes requested):
first round review done - except the models split refactor that was done post my review started.

imanadi commented on simple-app/src/test/resources/application.yml:
```diff
@@ -0,0 +1,9 @@
+      uri: mongodb://admin:password@host.docker.internal:37017/grayskull-test?authSource=admin
+grayskull:
+  crypto:
+    keys:
+      test-key: "MfDeylVH3ZmvnrK+XJv259fzjFB55+9yZaUr1OYvNVE=" 
```
Resolving the comment as the commited key is a test-key.

imanadi commented on server/pom.xml:
```diff
@@ -90,6 +96,42 @@
+                            <version>1.5.5.Final</version>
+                        </path>
+                        <path>
+                            <groupId>org.projectlombok</groupId>
+                            <artifactId>lombok</artifactId>
+                            <version>${lombok.version}</version>
+                        </path>
```
Resolved, as the parent pom.xml declares lombok.version. 
`<lombok.version>1.18.38</lombok.version>`

imanadi commented on server/src/main/java/com/flipkart/grayskull/controllers/AdminController.java:
```diff
@@ -0,0 +1,29 @@
+    @GetMapping("/project/{projectId}/secrets/{secretName}/data/{version}")
+    public SecretDataVersionResponse getSecretDataVersion(@PathVariable("projectId") @NotBlank @Size(max = 255) String projectId, @PathVariable("secretName") @NotBlank @Size(max = 255) String secretName, @PathVariable("version") @Min(1) int version) {
+        return secretService.getSecretDataVersion(projectId, secretName, version);
+    }
```
Resolving, as AdminController was deprecated.

imanadi commented on models/src/main/java/com/flipkart/grayskull/models/dto/response/SecretDataResponse.java:
```diff
@@ -0,0 +1,23 @@
+public class SecretDataResponse {
+    int dataVersion;
+    String publicPart;
+    String privatePart;
```
It's design is intended. 

 * Secret data values with version metadata.
 * Contains sensitive information for authorized access.
 
SecretMetadata dto provides non sensitive responses.

imanadi commented on models/src/main/java/com/flipkart/grayskull/models/exceptions/GrayskullException.java:
```diff
@@ -0,0 +1,47 @@
+    public GrayskullException(ErrorCode errorCode, String message, Throwable cause) {
+        super(message, cause);
+        this.errorCode = errorCode;
+    }
```
Resolving the outdated comment.

imanadi commented on models/src/main/java/com/flipkart/grayskull/models/dto/response/SecretDataVersionResponse.java:
```diff
@@ -0,0 +1,23 @@
+public class SecretDataVersionResponse {
+    int dataVersion;
+    String publicPart;
+    String privatePart;
```
Resolving the outdated comment.

imanadi commented on models/src/main/java/com/flipkart/grayskull/models/dto/response/CreateSecretResponse.java:
```diff
@@ -0,0 +1,22 @@
+@Data
+@AllArgsConstructor
+@JsonInclude(JsonInclude.Include.NON_NULL)
+public class CreateSecretResponse {
```
Resolving the outdated comment. It is also a duplicate comment which is addressed and resolved above.

imanadi commented on models/src/main/java/com/flipkart/grayskull/models/dto/response/SecretMetadata.java:
```diff
@@ -0,0 +1,26 @@
+@Data
+@AllArgsConstructor
+@JsonInclude(JsonInclude.Include.NON_NULL)
+public class SecretMetadata {
```
Resolving the outdated comment. It is also a duplicate comment which is addressed and resolved above.

imanadi commented on models/src/main/java/com/flipkart/grayskull/models/db/AuditEntry.java:
```diff
@@ -0,0 +1,68 @@
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+public class AuditEntry {
```
All DB models are designed to be database agnostic, which is why they don't have MongoDB-specific annotations.

imanadi commented on server/src/main/java/com/flipkart/grayskull/service/utils/SecretEncryptionUtil.java:
```diff
@@ -0,0 +1,30 @@
+
+import java.util.Base64;
+
+@Component
+@RequiredArgsConstructor
+public class SecretEncryptionUtil {
+
+    private final EncryptionService encryptionService;
+
+    public void encryptSecretData(SecretData secretData, String keyId) {
+        if (secretData.getPrivatePart() != null && !secretData.getPrivatePart().isEmpty()) {
+            byte[] encrypted = encryptionService.encrypt(secretData.getPrivatePart().getBytes(), keyId);
+            secretData.setPrivatePart(Base64.getEncoder().encodeToString(encrypted));
+            secretData.setKmsKeyId(keyId);
+        }
+    }
+
+    public void decryptSecretData(SecretData secretData) {
+        if (secretData.getPrivatePart() != null && !secretData.getPrivatePart().isEmpty()) {
+            byte[] decrypted = encryptionService.decrypt(Base64.getDecoder().decode(secretData.getPrivatePart()), secretData.getKmsKeyId());
+            secretData.setPrivatePart(new String(decrypted));
+        }
+    }
+} 
```
Resolving the suggestion as there isn't much duplicate code as highlighted by coderabbitai

imanadi commented on server/src/main/java/com/flipkart/grayskull/controllers/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,46 @@
+
+import com.flipkart.grayskull.models.dto.response.ErrorResponse;
+import com.flipkart.grayskull.models.exceptions.DuplicateSecretException;
+import com.flipkart.grayskull.models.exceptions.InvalidProjectConfigurationException;
```
Resolving the outdated comment.

imanadi commented on server/src/main/java/com/flipkart/grayskull/service/implementations/SecretServiceImpl.java:
```diff
@@ -0,0 +1,196 @@
+    public void deleteSecret(String projectId, String secretName) {
+        Secret secret = findSecretOrThrow(projectId, secretName);
+        secretRepository.delete(secret);
+    }
```
It is an outdated comment. 
The new code addresses this issue. 

`/**
     * Disables a secret, marking it as soft-deleted.
     *
     * @param projectId  The ID of the project.
     * @param secretName The name of the secret to disable.
     */
    @Override
    @Transactional
    @Auditable(action = AuditAction.DELETE_SECRET)
    public void deleteSecret(String projectId, String secretName) {
        Secret secret = findActiveSecretOrThrow(projectId, secretName);
        secret.setState(SecretState.DISABLED);
        secret.setUpdatedBy(authnUtil.getCurrentUsername());
        secretRepository.save(secret);
    }`

imanadi commented on models/src/main/java/com/flipkart/grayskull/models/enums/SecretProvider.java:
```diff
@@ -0,0 +1,16 @@
+    /**
+     * The secret is provided by the database credentials management system.
+     */
+    DBCREDS
```
It was discussed to provide and release support for limited set of SecretProvider, and to enforce this, it is ideal to have SecretProvider as enum than taking the input from the User.

imanadi commented on setup/setup.sh:
```diff
@@ -0,0 +1,133 @@
+        source .env
+    fi
+
+    log "INFO" "Initializing MongoDB replica set..."
+    if ! docker exec grayskull-local-mongo1 mongosh --username "$MONGO_ROOT_USERNAME" --password "$MONGO_ROOT_PASSWORD" --eval '
+        rs.initiate({
+            _id: "rs0",
+            members: [
+                {_id: 0, host: "grayskull-local-mongo1:27017", priority: 2},
+                {_id: 1, host: "grayskull-local-mongo2:27017", priority: 1},
+                {_id: 2, host: "grayskull-local-mongo3:27017", priority: 1}
+            ]
+        })
+    '; then
+        log "ERROR" "Failed to initialize replica set"
+        exit 1
+    fi
```
Resolving, as there is a teardown script to resolve such mishaps and it is not critical as the setup is meant for local setup.

imanadi commented on setup/.env:
```diff
@@ -0,0 +1,3 @@
+# MongoDB Configuration for Grayskull Local Development
+MONGO_ROOT_USERNAME=admin
+MONGO_ROOT_PASSWORD=password
```
Resolving the comment as the config is intended to be used for local setup, not production setup.

imanadi commented on server/src/test/java/com/flipkart/grayskull/BaseIntegrationTest.java:
```diff
@@ -0,0 +1,114 @@
+                .with(user(username)));
+    }
+
+    protected ResultActions performListSecrets(String projectId, String username, String... queryParams) throws Exception {
+        String url = "/v1/project/{projectId}/secrets";
+        if (queryParams.length > 0) {
+            url += "?" + String.join("&", queryParams);
+        }
+        return mockMvc.perform(get(url, projectId)
+                .with(user(username)));
+    }
```
Ignoring the AI's suggestion in the ITs.

imanadi commented on server/src/main/java/com/flipkart/grayskull/service/utils/AuthnUtil.java:
```diff
@@ -0,0 +1,20 @@
+     *
+     * @return The name of the current user.
+     */
+    public String getCurrentUsername() {
+        return SecurityContextHolder.getContext().getAuthentication().getName();
+    }
```
Acknowledging, will implement.

imanadi commented on setup/setup.sh:
```diff
@@ -0,0 +1,133 @@
+max_attempts=30
+attempt=1
+while [ $attempt -le $max_attempts ]; do
+    if docker exec grayskull-local-mongo1 mongosh --quiet --eval "1" >/dev/null 2>&1; then
+        break
+    fi
+    log "INFO" "Waiting for MongoDB to be ready (attempt $attempt/$max_attempts)..."
```
Resolving the outdated comment. It is also a duplicate comment which is addressed and resolved above.

imanadi commented on models/src/main/java/com/flipkart/grayskull/models/dto/request/UpgradeSecretDataRequest.java:
```diff
@@ -0,0 +1,26 @@
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+public class UpgradeSecretDataRequest {
```
I didn’t have a strong intention behind choosing one method prefix between upgrade and update. I opted for upgrade because it conveys the idea of improving the secret in a positive way, which aligns with the intent of the API. However, I don’t have a firm stance on this and can switch to update if provided with a logical reason.

imanadi commented on models/src/main/java/com/flipkart/grayskull/models/enums/AuditAction.java:
```diff
@@ -0,0 +1,19 @@
+package com.flipkart.grayskull.models.enums;
```
Acknowledging, I will think, propose, and discuss to refactor.

imanadi commented on spi/pom.xml:
```diff
@@ -18,6 +18,21 @@
     </properties>
 
     <dependencies>
+        <dependency>
+            <groupId>com.flipkart.grayskull</groupId>
+            <artifactId>models-api</artifactId>
+            <version>${project.version}</version>
+        </dependency>
+        <dependency>
+            <groupId>com.flipkart.grayskull</groupId>
+            <artifactId>models-db</artifactId>
+            <version>${project.version}</version>
+        </dependency>
```
It was intended. Will reconsider based on some already existing human comments.

imanadi commented on server/src/main/java/com/flipkart/grayskull/audit/utils/AuditMaskingSerializer.java:
```diff
@@ -0,0 +1,30 @@
+ */
+public class AuditMaskingSerializer extends JsonSerializer<Object> {
+
+    private static final String MASKED_VALUE = "***MASKED***";
```
Spring Boot: masks sensitive properties in logs as ****** (fixed-length asterisks).
To follow a somewhat similar standard, I opted for stars *. 
Would you happen to have some reason for the suggestion?
 
Reference: https://www.baeldung.com/logback-mask-sensitive-data

imanadi commented on server/src/main/java/com/flipkart/grayskull/audit/AuditAspect.java:
```diff
@@ -0,0 +1,177 @@
+    private void audit(JoinPoint joinPoint, Auditable auditable, Object result) {
+        Map<String, Object> arguments = getMethodArguments(joinPoint);
+
+        String projectId = (String) arguments.getOrDefault("projectId", "UNKNOWN");
```
Acknowledged. I will create a constant and use it.

imanadi commented on models/src/main/java/com/flipkart/grayskull/models/db/AuditEntry.java:
```diff
@@ -0,0 +1,68 @@
+     * The name of the {@link Secret} related to this audit event, if applicable.
+     * Combined with {@code projectId} and {@code secretVersion}, this precisely identifies the subject of the audit.
+     */
+    private String secretName;
```
Acknowledged, I will redesign to extend.

imanadi commented on models/src/main/java/com/flipkart/grayskull/models/dto/request/UpgradeSecretDataRequest.java:
```diff
@@ -0,0 +1,19 @@
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+public class UpgradeSecretDataRequest {
```
Resolving the comment, Implemented changes in commit: https://github.com/flipkart-incubator/grayskull/pull/3/commits/d9cf1a9062b842ffaa8e3d8393eff2e92ecac689

imanadi commented on server/src/main/java/com/flipkart/grayskull/audit/AuditAspect.java:
```diff
@@ -0,0 +1,177 @@
+     * @return The secret name, or "UNKNOWN" if not found.
+     */
+    private String extractSecretName(JoinPoint joinPoint, Map<String, Object> arguments) {
+        Object name = arguments.get("secretName");
```
Resolving the comment, Implemented changes in commit: https://github.com/flipkart-incubator/grayskull/pull/3/commits/d9cf1a9062b842ffaa8e3d8393eff2e92ecac689

imanadi commented on server/src/main/java/com/flipkart/grayskull/service/impl/SecretServiceImpl.java:
Resolving the comment, Implemented changes in commit: https://github.com/flipkart-incubator/grayskull/pull/3/commits/d9cf1a9062b842ffaa8e3d8393eff2e92ecac689

imanadi commented on server/src/main/java/com/flipkart/grayskull/audit/AuditAspect.java:
```diff
@@ -0,0 +1,177 @@
+     * @return A map of metadata for the audit entry.
+     */
+    private Map<String, String> buildMetadata(Map<String, Object> arguments, Object result) {
+        Map<String, String> metadata = new HashMap<>();
+        arguments.forEach((key, value) -> {
+            if (value != null) {
+                try {
+                    metadata.put(key, OBJECT_MAPPER.writeValueAsString(value));
+                } catch (JsonProcessingException e) {
+                    metadata.put(key, "Error serializing object: " + e.getMessage());
+                }
+            }
+        });
+
+        if (result != null) {
+            try {
+                metadata.put("result", OBJECT_MAPPER.writeValueAsString(result));
+            } catch (JsonProcessingException e) {
+                metadata.put("result", "Error serializing object: " + e.getMessage());
+            }
+        }
```
Acknowledged, will propose a better structure.

imanadi replied:
After analysis, I am still in favor of the current Audit Entity design for its top-level fields.
While fields such as projectId, resourceType, and resourceName may be duplicated in the metadata, they help maintain proper query structure and performance. The main purpose of auditing is to answer "who did what on which object" - the "which" part is clearly identifiable thanks to these top-level indexed fields.

{
  "_id": ObjectId('68ef8e9b6dd74761ed4bf7ff'),
  "projectId": "project",           // ✅ Keep -  identifies the resource
  "resourceType": "SECRET",          // ✅ Keep - enables filtering by type
  "resourceName": "secret-1",        // ✅ Keep - identifies the resource
  "resourceVersion": 1,              // ✅ Keep - tracks version
  "action": "CREATE_SECRET",         // ✅ Keep - what was done
  "userId": "user1",                 // ✅ Keep - who did it
  "timestamp": ISODate('2025-10-15T12:07:55.076Z'),
  "_class": "com.flipkart.grayskull.entities.AuditEntryEntity"
}

The metadata field stores the request and response metadata.

grssam replied:
The original issue around request and response parameters remains in the latest commit. the entire response object is put in the key `RESULT_METADATA_KEY` while individual request params are added directly.. rather than some key like `REQUEST_METADATA_KEY`

imanadi commented on server/src/main/java/com/flipkart/grayskull/authz/GrayskullSecurity.java:
```diff
@@ -0,0 +1,84 @@
+                            // Secret does not exist, fall back to a project-level check.
+                            AuthorizationContext context = AuthorizationContext.forProject(authentication, project);
+                            return authorizationProvider.isAuthorized(context, action);
+                        }))
```
**Answer**: YES, the authorization model DOES support both project-level and secret-level permissions, and the current implementation handles this correctly.

**How it works:**
1. Authorization rules can specify an optional `secret` field (see `AuthorizationProperties.Rule`)
2. Rules WITHOUT a `secret` field are project-level rules that apply to all secrets in the project
3. Rules WITH a `secret` field are secret-level rules that apply only to specific secrets

**Why the current fallback is correct:**
The fallback to project-level check in `GrayskullSecurity.hasPermission(projectId, secretName, action)` (lines 89-92) when a secret doesn't exist is appropriate because:

1. When we fall back, we create `AuthorizationContext.forProject(authentication, project)` - a context WITHOUT a secret
2. The authorization provider's `secretMatches()` method evaluates this correctly:
   - Project-level rules (secret field is null) will match because they're designed to apply to any secret context (line 55-56 in SimpleAuthorizationProvider)
   - Secret-level rules (secret field is specified) won't match because the rule requires a specific secret name

**Example scenario - works correctly:**
```yaml
rules:
  - user: alice
    project: "proj-1"
    actions: ["secrets.list"]  # Project-level: can list secrets
  
  - user: alice  
    project: "proj-1"
    secret: "db-password"  # Secret-level: can read ONLY this specific secret
    actions: ["secrets.read.value"]
```

When Alice tries to read a non-existent secret "api-key":
- Authorization falls back to project-level check
- Her project-level rule (without secret field) does NOT grant "secrets.read.value"
- Her secret-level rule (with secret="db-password") doesn't match because secret name differs
- Result: ❌ DENIED (correct behavior)

**The authorization provider controls the logic:**
As suggested in the comment, the behavior IS controlled by the authorization provider implementation. The `GrayskullSecurity` layer simply constructs the appropriate context (with or without secret) and delegates to the provider's `isAuthorized()` method. The provider's `secretMatches()` logic determines which rules apply.

**Conclusion**: No changes needed. The current implementation correctly delegates authorization decisions to the provider, which properly handles both project-level and secret-level permissions through its rule matching logic.

grssam replied:
While the example you've explained and the rationale is correct, how do you model mutually exclusive roles using this behavior? We want to follow the rule of minimum access, so one role can't just have every permission in a single role.
The default authz simple impl is ABAC and not RBAC - I think this is the main difference in opinion here. Think if RBAC model will fail here based on the logic in the above `hasPermission` method

grssam replied:
@imanadi  - please also create a github issue with the details around this for taking this up as a followup

imanadi replied:
Created an Issue with details and reference to the conversation: https://github.com/flipkart-incubator/grayskull/issues/4

imanadi commented on server/src/main/java/com/flipkart/grayskull/service/impl/SecretServiceImpl.java:
```diff
@@ -0,0 +1,245 @@
+    @Transactional
+    @Auditable(action = AuditAction.CREATE_SECRET)
+    public CreateSecretResponse createSecret(String projectId, CreateSecretRequest request) {
+        secretRepository.findByProjectIdAndName(projectId, request.getName())
```
Added in the latest commit: https://github.com/flipkart-incubator/grayskull/pull/3/commits/7fc33ee53b867c82c04ef7f1015ab97db6035303

imanadi commented on server/src/main/java/com/flipkart/grayskull/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,136 @@
+     * @return A ResponseEntity containing a standardized error response with a 400 Bad Request status.
+     */
+    @ExceptionHandler(MethodArgumentTypeMismatchException.class)
+    public ResponseEntity<ResponseTemplate<Void>> handleMethodArgumentTypeMismatchException(MethodArgumentTypeMismatchException ex, WebRequest request) {
```
Clubbed in the latest commit: https://github.com/flipkart-incubator/grayskull/pull/3/commits/7fc33ee53b867c82c04ef7f1015ab97db6035303

ssrangisetti commented on server/src/main/java/com/flipkart/grayskull/entities/SecretEntity.java:
```diff
@@ -0,0 +1,63 @@
+
+    @Id
+    @Override
+    public String getId() {
```
for secret also Id is an auto generated field, so lets move it to Entity from the Secret

ssrangisetti commented on server/src/main/java/com/flipkart/grayskull/models/dto/request/SecretDataBase.java:
```diff
@@ -0,0 +1,34 @@
+     * Public part (non-sensitive).
+     */
+    @NotBlank
+    private String publicPart;
```
we need limits on public and private parts. set some max length limit like 10,000

imanadi replied:
Added in https://github.com/flipkart-incubator/grayskull/pull/3/commits/9008e599b17253b8839ed2c0c938cb7c0cdca83f

ssrangisetti commented on server/src/main/java/com/flipkart/grayskull/service/impl/SecretServiceImpl.java:
```diff
@@ -0,0 +1,279 @@
+     * @return The existing or newly created {@link Project}.
+     */
+    @Transactional
+    public Project getOrCreateProject(String projectId) {
```
this Transactional annotation is not used and not needed. its not used because spring annotations don't work when called from same class. not needed because this is called from create and upgrade both of which are already in transaction. So this can be removed

imanadi commented on server/src/main/java/com/flipkart/grayskull/spimpl/repositories/AuditEntryRepositoryImpl.java:
```diff
@@ -0,0 +1,37 @@
+/**
+ * Internal MongoDB repository for AuditEntryEntity.
+ */
+interface AuditEntryMongoRepository extends MongoRepository<AuditEntryEntity, String> {
```
Made the requested changes in https://github.com/flipkart-incubator/grayskull/pull/3/commits/9008e599b17253b8839ed2c0c938cb7c0cdca83f

grssam commented on server/src/main/java/com/flipkart/grayskull/configuration/DefaultProjectConfig.java:
```diff
@@ -0,0 +1,26 @@
+@Setter
+public class DefaultProjectConfig {
+
+    private Project defaultProject;
```
can you ack here? looks like instead of this, we just need `defaultKMSId`

grssam commented on spi/pom.xml:
```diff
@@ -18,6 +17,21 @@
+        <dependency>
+            <groupId>org.springframework.data</groupId>
+            <artifactId>spring-data-commons</artifactId>
+            <scope>provided</scope>
```
ideally, no other pom than the root pom should even have the `<version>` tag for the dependencies. Also, why is this `provided` scope?

grssam replied:
How is 
``` 
           <groupId>org.springframework.data</groupId>
            <artifactId>spring-data-commons</artifactId>
```
an SPI module? and btw, interfaces are needed at runtime as well, not just at compile time.

imanadi replied:
`spring-data-commons` is not in the SPI module -(was removed). The SPI only has `jakarta.servlet-api`, `spring-security-core`, and `lombok` as `provided` dependencies. 

You're correct that interfaces are needed at runtime! The `provided` scope means: "I need these to compile my interface definitions, but I won't bundle them - the consuming application will provide them." The `server` module includes these same dependencies with `compile` scope, making them available at runtime. This prevents the SPI from forcing its dependency versions on consumers while still allowing it to reference framework types in its contract definitions. This was suggested by Sai to have it like this.

grssam commented on server/src/main/java/com/flipkart/grayskull/controllers/SecretController.java:
```diff
@@ -0,0 +1,101 @@
+import java.util.Optional;
+
+@RestController
+@RequestMapping("/v1/project/{projectId}/secrets")
```
plurals. `.../projects/...`

grssam replied:
This just means that it was a miss in the contract proposal review. Please fix this.

imanadi commented on server/src/main/java/com/flipkart/grayskull/entities/AuditEntryEntity.java:
```diff
@@ -0,0 +1,35 @@
+
+    @Id
+    @Override
+    public String getId() {
```
Why `getId()` and `getTimestamp()` are overridden in `AuditEntryEntity`

**The Problem:** Spring Data MongoDB needs `@Id` and `@CreatedDate` annotations to function, but the parent `AuditEntry` class in the `spi` module is a framework-agnostic POJO with no persistence dependencies.

**The Solution:** Override the getter methods in `AuditEntryEntity` to apply Spring Data annotations:

```java
@Id
@Override
public String getId() {
    return super.getId();
}
```

**Why This Matters:**
- **SPI stays clean:** The `spi` module has zero dependencies on Spring/MongoDB, making it reusable across different implementations
- **Annotations on getters:** Spring Data requires annotations on methods, not inherited fields
- **Separation of concerns:** The POJO defines the contract; the entity adds persistence behavior

**Note:** Discussed internally. As of now, keeping the same. ID can be moved to the internal entity, but as of now it is breaking functionality, so keeping the change same and overriding the method in the implementation keeping SPI intact as a plain POJO.

imanadi replied:
Created an Issue with details and reference to the conversation: https://github.com/flipkart-incubator/grayskull/issues/5

imanadi commented on server/src/main/java/com/flipkart/grayskull/audit/AuditAspect.java:
```diff
@@ -0,0 +1,192 @@
+     * @return The resource name, or "UNKNOWN" if not found.
+     */
+    private String extractResourceName(JoinPoint joinPoint, Map<String, Object> arguments) {
+        Object name = arguments.get(SECRET_NAME_PARAM);
```
Updated to extract resource name from response entities (source of truth for operations that return data), with fallback to method arguments for void operations like delete. Added `projectId` and `name` fields to `UpgradeSecretDataResponse` for consistency with `CreateSecretResponse` and to enable clean audit extraction from response schema. This hybrid approach ensures all operations are properly audited while following the principle that response schemas are the contract, not method parameters.

imanadi commented on server/src/main/java/com/flipkart/grayskull/service/impl/SecretServiceImpl.java:
```diff
@@ -0,0 +1,245 @@
+    public CreateSecretResponse createSecret(String projectId, CreateSecretRequest request) {
+        secretRepository.findByProjectIdAndName(projectId, request.getName())
+                .ifPresent(s -> {
+                    throw new ResponseStatusException(HttpStatus.CONFLICT, "A secret with the same name " + request.getName() + " already exists.");
```
This was implemented in previous commits.

imanadi commented on server/src/main/java/com/flipkart/grayskull/service/utils/AuthnUtil.java:
```diff
@@ -0,0 +1,20 @@
+     * @return The name of the current user.
+     */
+    public String getCurrentUsername() {
+        return SecurityContextHolder.getContext().getAuthentication().getName();
```
Module restructuring was already done previously.

grssam commented on server/src/main/java/com/flipkart/grayskull/audit/AuditAspect.java:
```diff
@@ -0,0 +1,192 @@
+     * @return The resource version, or {@code null} if not applicable.
+     */
+    private Integer extractResourceVersion(Object result) {
+        if (result instanceof CreateSecretResponse) {
```
you should rely on request object rather than response.
Also, let's rename the object names based on content rather than action here as well.. `CreateSecretResponse` -> `SecretResponse`

grssam replied:
@imanadi is still an issue. can you fix this? should be straight forward.
